tests/v2/validation/provisioning/rke2/etcd_bkp_restore_test.go

```diff
@@ -0,0 +1,385 @@
+					snapshotToBeRestored = snapshot.Name
+				}
+				totalClusterSnapShots++
+			}
+		}
+		if totalClusterSnapShots == 3 {
+			return true, nil
+		}
+		return false, nil
+	})
+	require.NoError(r.T(), err)
+
+	logrus.Infof("creating watch over cluster after getting the snapshots.............")
+	r.watchAndWaitForCluster(client, clusterName)
+	logrus.Infof("cluster is active again.............")
+
+	logrus.Infof("creating a workload(w2, deployment).............")
+	w2labels := map[string]string{}
+	w2labels["workload.user.cattle.io/workloadselector"] = fmt.Sprintf("apps.deployment-%v-%v", r.ns, w2Name)
+	containerTemplate2 := workloads.NewContainer("ngnix", "nginx", v1.PullAlways, []v1.VolumeMount{}, []v1.EnvFromSource{})
+	podTemplate2 := workloads.NewPodTemplate([]v1.Container{containerTemplate2}, []v1.Volume{}, []v1.LocalObjectReference{}, w2labels)
+	deployment2 := workloads.NewDeploymentTemplate(w2Name, r.ns, podTemplate2, w2labels)
+
+	steveclient, err = client.Steve.ProxyDownstream(clusterID)
+	require.NoError(r.T(), err)
+
+	deploymentResp2, err := steveclient.SteveType(workloads.DeploymentSteveType).Create(deployment2)
+	require.NoError(r.T(), err)
+
+	require.Equal(r.T(), w2Name, deploymentResp2.ObjectMeta.Name)
+	logrus.Infof("created a workload(w2, deployment).............")
+
+	logrus.Infof("creating watch over w2.............")
+	r.watchAndWaitForDeployment(client, clusterID, w2Name)
+	logrus.Infof("w2 is ready.............")
+
+	logrus.Infof("upgrading cluster k8s version.............")
+	k8sUpgradedVersion := r.kubernetesVersions[1]
+	r.upgradeClusterK8sVersion(client, clusterName, k8sUpgradedVersion)
+	r.watchAndWaitForCluster(client, clusterName)
+	logrus.Infof("cluster is active again.............")
+
+	cluster, _, err := getProvisioningClusterByName(client, clusterName, r.ns)
+	require.NoError(r.T(), err)
+	require.Equal(r.T(), k8sUpgradedVersion, cluster.Spec.KubernetesVersion)
+
+	logrus.Infof("restoring snapshot.............")
+	require.NoError(r.T(), restoreSnapshot(client, clusterName, snapshotToBeRestored, 1, "all", r.ns))
+	logrus.Infof("successfully submitted restoration request.............")
+
+	logrus.Infof("creating watch over cluster after restore.............")
+	r.watchAndWaitForCluster(client, clusterName)
+	logrus.Infof("cluster is active again.............")
+
+	logrus.Infof("creating watch over pods.............")
+	r.watchAndWaitForPods(client, clusterID)
+	logrus.Infof("All pods are up and running.............")
+
+	logrus.Infof("fetching deployment list to validate restore.............")
+	deploymentsList, err := deployments.ListDeployments(client, clusterID, &types.ListOpts{
+		Filters: map[string]interface{}{
+			"fieldSelector": "metadata.namespace=" + r.ns,
+		},
+	})
+	require.NoError(r.T(), err)
+	require.Equal(r.T(), 1, len(deploymentsList.Items))
+	require.Equal(r.T(), w1Name, deploymentsList.Items[0].ObjectMeta.Name)
+	logrus.Infof(" deployment list validated successfully.............")
```

Comment:
Rishab I know this a nitpick, but after discussing with Ugur offline, I don't think having the ListDeployments or ListIngresses as extensions is necessary, especially since we wouldn't want to convert all the time. I don't want to set the precedence with Steve list calls.

For example for above why not just do?
deployments, err := steveclient.SteveType(workloads.DeploymentSteveType).List()

deployments.Data[0].ObjectMeta.Name
and same with ingresses, why is the whole conversion needed? I want to remind you the Steve response is slightly different from the conversion, (that is how the API works), so why is the conversion necessary. Also why are you using list rather than ByID? Wouldn't make more sense to use ByID?

Reply:
As we've discussed on this, I totally agree with this.

I think steve's client is capable enough to list (or any CRUD) status, spec, or the resource object via JSONResp without any extension need. Dynamic client conversion wasn't clear and long enough therefore the extensions centered around the dynamic client were required. So I don't think ListIngress and ListDeployments are necessary.

As Israel mentioned, since the tests are using the first element in the list which is filtered by deployment name, steve's client `ByID` might be easier to reach that struct.

Reply:
Okay, removed the ListDeployment and ListIngress helper functions.